File: Course-Registration-System-main/java-(iteration_1)/src/StudentController.java
import java.util.ArrayList;
import java.util.Scanner;

public class StudentController {
    private ArrayList<Course> courses;
    Student student;

    public StudentController(Student student, ArrayList<Course> courses) {
        this.student = student;
        this.courses = courses;
    }

    public String[] takeCoursesInput() {
        Scanner scanner = new Scanner(System.in);
        Scanner scanner1 = new Scanner(System.in);
        System.out.print("How many courses do you want to enroll?(Please type integer!): ");
        String input = scanner.nextLine();
        String[] courseSectionCodes = new String[10];
        int intValue;
        try {
            // Attempt to parse the input as an integer
            intValue = Integer.parseInt(input);
            // If parsing is successful, it's an integer

        } catch (NumberFormatException e) {
            // If an exception is caught, it's not an integer
            intValue = 0;
            System.out.println("Error: Input is not an integer!");
        }

        for (int i = 0; i < intValue; i++) {
            String code;
            System.out.print("Enter the " + (i + 1) + ". course code with section no (e.g: CSE2225.1): ");
            code = scanner1.nextLine();
            courseSectionCodes[i] = code;
        }

        return courseSectionCodes;
    }

    public static void printCourseSchedule(ArrayList<CourseSection> sections) {
        System.out.println("Course Schedule:");
        System.out.printf("%-15s%-15s%-15s\n",
                "Classroom", "Day", "Hour");
        System.out.println("-------------------------");

        for (CourseSection section : sections) {
            System.out.printf("%-15s%-15s%-15s\n",
                    section.getClassroom(), section.getDay(), section.getHour());
        }
    }

    public void showOptions() {
        Scanner input = new Scanner(System.in);
        String selection;

        do {
            System.out.println("Type [1] Enroll a course");
            System.out.println("Type [2] List enrolled courses");
            System.out.println("Type [3] View Transcript");
            System.out.println("Type [4] View Info");
            System.out.println("Type [5] Requested courses info");
            System.out.println("Type [6] View Time Table");
            System.out.println("Type [q] to quit");
            System.out.print("-> ");
            selection = input.nextLine();

            switch (selection) {
                case "1":

                    RegistrationControl r = new RegistrationControl(courses);
                    ArrayList<Course> kurs = r.getAvailableCourses(student.getTranscript(), student.getSemester());
                    for (int i = 0; i < kurs.size(); i++) {
                        for (int j = 0; j < kurs.get(i).getCourseSections().size(); j++) {
                            System.out.println(kurs.get(i).getCourseSections().get(j).viewCourseSectionInfo());
                        }
                    }
                    String requestedCourseSectionCode[] = takeCoursesInput();
                    CourseSection a = null;
                    for (int j = 0; j < requestedCourseSectionCode.length; j++) {
                        a = null;
                        if (requestedCourseSectionCode[j] != null) {
                            if (r.isCourseSectionCodeValid(kurs, requestedCourseSectionCode[j])) {
                                a = r.getCourseSection(kurs, requestedCourseSectionCode[j]);
                                if ((r.isHasCapacity(a))) {
                                    if (!(r.isOverlapping(student.getCurrentCourses(), a))) {
                                        if (r.currentTotalCredits(student.getCurrentCourses()) < 30) {
                                            CourseRequest request = new CourseRequest(student.getId(), a, "waiting");
                                            System.out.println(a.getCourseSectionCode() + " successfully requested.");
                                            if (student.getCourseRequests() == null) {
                                                ArrayList<CourseRequest> temp = student.getCourseRequests();
                                                temp = new ArrayList<>();
                                                temp.add(request);
                                                student.setCourseRequests(temp);

                                            } else
                                                student.getCourseRequests().add(request);

                                        } else
                                            System.out.println("Failed: Total credits taken should not exceed 30.0!");
                                    } else
                                        System.out.println("Failed: " + a.getCourseSectionCode() + " overlapping!");
                                } else
                                    System.out.println("Failed: " + a.getCourseSectionCode() + " section is full!");
                            } else
                                System.out.println("Failed: Section code is not valid!");
                        }
                    }

                    break;
                case "2":
                    for (int i = 0; i < student.listCurrentCourses().length; i++)
                        if (student.listCurrentCourses()[i] != null) {
                            System.out.println(student.listCurrentCourses()[i]);
                        }
                    break;
                case "3":
                    for (int j = 0; j < student.getTranscript().viewTranscript().length; j++)
                        if (student.getTranscript().viewTranscript()[j] != null) {
                            System.out.println(student.getTranscript().viewTranscript()[j]);
                        }
                    break;
                case "4":
                    System.out.println(student.getInfo());
                    break;
                case "5":
                    if (student.getCourseRequests() != null) {
                        for (int i = 0; i < student.getCourseRequests().size(); i++) {
                            System.out.println(student.getCourseRequests().get(i).viewRequestInfo());
                        }
                    } else {
                        System.out.println("There is no course request");
                    }
                    break;
                case "6":
                    printCourseSchedule(student.getCurrentCourses());
                    break;
                case "q":
                    TranscriptSaver transcriptSaver = new TranscriptSaver();
                    transcriptSaver.save(student.getTranscript());
                    break;
                default:
                    System.out.println("Wrong input! Please enter 1,2,3,4 or q");
            }
        } while (!selection.equals("q"));

    }
}

File: Course-Registration-System-main/python-(iteration_2)/src/StudentController.py
from RegistrationControl import RegistrationControl
from TranscriptSaver import TranscriptSaver
from CourseRequest import CourseRequest
from CourseSection import CourseSection
import logging

class StudentController:
    def __init__(self, student, courses):
        self.__student = student
        self.__courses = courses
        self.logger = logging.getLogger(__name__)
        logging.basicConfig(level=logging.INFO)

    def takeCoursesInput(self):
        courseSectionCodes = []
        try:
            numCourses = int(input("How many courses do you want to enroll? (Please type an integer): "))
        except ValueError:
            print("Error: Input is not an integer!")
            return []

        for i in range(numCourses):
            code = input(f"Enter the {i + 1}. course code with section no (e.g., CSE2225.1): ")
            courseSectionCodes.append(code)

        return courseSectionCodes

    @staticmethod
    def printCourseSchedule(sections):
        print("Course Schedule:")
        print("{:<15}{:<15}{:<15}".format("Classroom", "Day", "Hour"))
        print("-------------------------")

        for section in sections:
            print("{:<15}{:<15}{:<15}".format(section.getClassroom(), section.getDay(), section.getHour()))

    def showOptions(self):
        selection = ""
        while selection != "q":
            print("Type [1] Enroll a course")
            print("Type [2] List enrolled courses")
            print("Type [3] View Transcript")
            print("Type [4] View Info")
            print("Type [5] Requested courses info")
            print("Type [6] View Time Table")
            print("Type [q] to quit")
            selection = input("-> ")

            if selection == "1":
                self.enrollCourse()
            elif selection == "2":
                self.listEnrolledCourses()
            elif selection == "3":
                transcript = self.viewTranscript()

            elif selection == "4":
                print(self.__student.getInfo())
            elif selection == "5":
                self.viewRequestedCourses()
            elif selection == "6":
                self.printCourseSchedule(self.__student.getCurrentCourses())
            elif selection == "q":
                transcriptSaver = TranscriptSaver()
                transcriptSaver.save(self.__student.getTranscript())
            else:
                print("Wrong input! Please enter 1, 2, 3, 4, 5, 6, or q")

    def enrollCourse(self):
        registrationControl = RegistrationControl(self.__courses)
        availableCourses = registrationControl.getAvailableCourses(self.__student.getTranscript(), self.__student.getSemester())

        for course in availableCourses:
            for section in course.getCourseSections():
                print(section.viewCourseSectionInfo())

        requestedCourseSectionCodes = self.takeCoursesInput()

        for code in requestedCourseSectionCodes:
            if code is not None:
                if registrationControl.isCourseSectionCodeValid(availableCourses, code):
                    courseSection = registrationControl.getCourseSection(availableCourses, code)
                    if registrationControl.hasCapacity(courseSection):
                        if not registrationControl.isOverlapping(self.__student.getCurrentCourses(), courseSection):
                            if registrationControl.currentTotalCredits(self.__student.getCurrentCourses()) < 30:

                                courseRequest = CourseRequest(self.__student.getId(), courseSection, "waiting")
                                self.logger.info("Request for %s set as 'waiting' for student with ID: %d", courseRequest.getCourseSectionCode(), self.__student.getId())

                                if self.__student.getCourseRequests() is None:
                                    self.__student.setCourseRequests([courseRequest])
                                else:
                                    self.__student.getCourseRequests().append(courseRequest)
                            else:
                                print("Failed: Total credits taken should not exceed 30.0!")
                        else:
                            print(f"Failed: the course is overlapping!")
                    else:
                        print(f"Failed: the course section is full!")
                else:
                    print("Failed: Section code is not valid!")

    def listEnrolledCourses(self):
        for courseSection in self.__student.listCurrentCourses():
            if courseSection is not None:
                print(courseSection)

    def viewTranscript(self):
        for grade in self.__student.getTranscript().viewTranscript():
            if grade is not None:
                print(grade)

    def viewRequestedCourses(self):
        if self.__student.getCourseRequests() is not None:
            for request in self.__student.getCourseRequests():
                print(request.viewRequestInfo())
        else:
            print("There is no course request")
